tokio-util/src/sync/cancellation_token.rs

```diff
@@ -222,3 +254,107 @@ impl<'a> Future for WaitForCancellationFuture<'a> {
         }
     }
 }
+
+// ===== impl WaitForCancellationFutureOwned =====
+
+impl core::fmt::Debug for WaitForCancellationFutureOwned {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("WaitForCancellationFutureOwned").finish()
+    }
+}
+
+impl Future for WaitForCancellationFutureOwned {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+        let this = unsafe { Pin::into_inner_unchecked(self) };
+
+        loop {
+            if this.cancellation_token.is_cancelled() {
+                return Poll::Ready(());
+            }
+
+            let future = this
+                .get_future_mut()
+                // Safety:
+                //
+                // `self` is pinned, so self.future is also pinned.
+                .map(|fut| unsafe { Pin::new_unchecked(fut) });
+
+            // No wakeups can be lost here because there is always a call to
+            // `is_cancelled` between the creation of the future and the call to
+            // `poll`, and the code that sets the cancelled flag does so before
+            // waking the `Notified`.
+            if future
+                .map(|fut| fut.poll(cx).is_pending())
+                .unwrap_or_default()
+            {
+                return Poll::Pending;
+            }
+
+            // self.future is either None, or a future that is already
+            // polled and returns `Poll::Ready`, which should not be
+            // polled again.
+            this.do_drop_future();
+
+            let notified = this.cancellation_token.inner.notified();
+
+            // Safety:
+            //  - We transmute notified into Notified<'static>
+            //    since there's no 'self
+            //  - self is pinned, so it's safe to have self-reference
+            //    data structure.
+            //  - this.cancellation_token contains an Arc to TreeNode,
+            //    which is guaranteed to have stable dereference.
+            //    So even if `Notified` implements `Unpin` and `self`
+            //    get moved, it would still be valid.
+            //  - We never use Notified<'static>.
+            //    Instead, it is transmuted back to Notified<'a>
+            //    where 'a is lifetime of self before being used.
+            let notified: tokio::sync::futures::Notified<'static> =
+                unsafe { mem::transmute(notified) };
+
+            this.future = Some(mem::ManuallyDrop::new(notified));
+        }
+    }
+}
+
+impl WaitForCancellationFutureOwned {
+    fn do_drop_future<'a>(&'a mut self) {
+        if let Some(future) = self.future.as_mut() {
+            // Safety:
+            //
+            // The future itself refererences cancellation_token, so its
+            // lifetime must be at least as long as 'a.
+            let future: &mut mem::ManuallyDrop<tokio::sync::futures::Notified<'a>> =
+                unsafe { mem::transmute(future) };
```

Comment:
You don't have to do this transmute. You can run the destructor on a `Notified<'static>`.

Reply:
[`Notified<'static>::drop`](https://docs.rs/tokio/1.21.2/src/tokio/sync/notify.rs.html#900-942) dereferences the `&'static Notify` to access `Notify::state`.

Without the transmute, I think miri would report this as an error since we are creating `'static` lifetime out of a `'a` lifetime which has smaller scope.

Reply:
Thanks for the explanation, I've removed this `mem::transmute`.

Reply:
BTW, I've also removed the other `mem::transmute` used in similar way to produce `Option<&mut Notified<'a>>`.